File: android/app/src/main/java/tech/numberless/port/FileUploaders/RegularUploader.kt
package tech.numberless.port.fileuploaders

import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import io.ktor.client.HttpClient
import io.ktor.client.engine.cio.CIO
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.request.forms.formData
import io.ktor.client.request.forms.submitFormWithBinaryData
import io.ktor.client.request.post
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.Headers
import io.ktor.http.HttpHeaders
import io.ktor.http.isSuccess
import io.ktor.serialization.kotlinx.json.json
import kotlinx.serialization.json.Json
import tech.numberless.port.fileuploaders.SingleShotUploader
import java.io.File

/**
 * Data class representing the fields for a non-multipart upload
 * @param key The key for the upload
 * @param policy The policy for the upload
 * @param xAmzAlgorithm The x-amz-algorithm for the upload
 * @param xAmzCredential The x-amz-credential for the upload
 * @param xAmzDate The x-amz-date for the upload
 * @param xAmzSecurityToken The x-amz-security-token for the upload
 * @param xAmzSignature The x-amz-signature for the upload
 */
data class SingleUploadFields(
    val key: String,
    val policy: String,
    @SerializedName("x-amz-algorithm") val xAmzAlgorithm: String,
    @SerializedName("x-amz-credential") val xAmzCredential: String,
    @SerializedName("x-amz-date") val xAmzDate: String,
    @SerializedName("x-amz-security-token") val xAmzSecurityToken: String,
    @SerializedName("x-amz-signature") val xAmzSignature: String,
)

/**
 * Data class representing the response for a non-multipart upload URL request
 * @param statusCode The status code of the response
 * @param body The body of the response
 */
data class SingleUploadURLResponse(
    val statusCode: Int,
    val body: SingleUploadBody,
)

/**
 * Data class representing the body of a non-multipart upload URL response
 * @param url The URL to upload to
 * @param mediaId The media ID
 */
data class SingleUploadBody(
    val url: SingleUploadURLData,
    val mediaId: String,
)

/**
 * Data class representing the URL and fields for a non-multipart upload
 * @param url The URL to upload to
 * @param fields The fields for the upload
 */
data class SingleUploadURLData(
    val url: String,
    val fields: SingleUploadFields,
)

class SingleShotUploader constructor(
    private val presignedFetchResource: String,
) : IUploader {
    override val type = "Single-shot"

    /**
     * Uploads a file to a presigned URL - non-multipart
     * @param client Ktor client
     * @param token Ignored
     * @param path Path to the file to upload
     * @return Media ID
     */
    override suspend fun upload(
        token: String, // Ignored by the backend, in this context
        path: String,
    ): String {
        HttpClient(CIO) {
            install(ContentNegotiation) {
                json(
                    Json {
                        prettyPrint = true
                        isLenient = true
                    },
                )
            }
        }.use { client ->
            val file =
                File(path).takeIf { it.exists() }
                    ?: throw IllegalArgumentException("File not found: $path")

            val res = client.post(presignedFetchResource)
            if (!res.status.isSuccess()) {
                throw Exception("Fetching presigned URL failed with response ${res.status.value}")
            }
            val (mediaId, urlData) =
                Gson()
                    .fromJson(res.bodyAsText(), SingleUploadURLResponse::class.java)
                    .body
                    .let { it.mediaId to it.url }

            if (!client
                    .submitFormWithBinaryData(
                        url = urlData.url,
                        formData =
                            formData {
                                // Add all fields from UrlData
                                mapOf(
                                    "key" to urlData.fields.key,
                                    "x-amz-algorithm" to urlData.fields.xAmzAlgorithm,
                                    "x-amz-credential" to urlData.fields.xAmzCredential,
                                    "x-amz-date" to urlData.fields.xAmzDate,
                                    "x-amz-security-token" to urlData.fields.xAmzSecurityToken,
                                    "policy" to urlData.fields.policy,
                                    "x-amz-signature" to urlData.fields.xAmzSignature,
                                ).forEach { (key, value) -> append(key, value) }
                                // Add the file
                                append(
                                    "file",
                                    file.readBytes(),
                                    Headers.build {
                                        append(HttpHeaders.ContentType, "application/octet-stream")
                                        append(HttpHeaders.ContentDisposition, "filename=\"${file.name}\"")
                                    },
                                )
                            },
                    ).status
                    .isSuccess()
            ) {
                throw Exception("Uploading file failed")
            }
            return mediaId
        }
    }
}

File: ios/Port/PortMediaUploader/FileUploaders/RegularUploader.swift
//
//  RegularUploader.swift
//  Port
//
//  Created by Shantanav Saurav on 18/04/2025.
//

import Foundation

// Expected JSON structure for a single-upload response.
struct URLData: Decodable {
  struct Fields: Decodable {
    let key: String
    let xAmzAlgorithm: String
    let xAmzCredential: String
    let xAmzDate: String
    let xAmzSecurityToken: String
    let policy: String
    let xAmzSignature: String
    enum CodingKeys: String, CodingKey {
      case key
      case xAmzAlgorithm = "x-amz-algorithm"
      case xAmzCredential = "x-amz-credential"
      case xAmzDate = "x-amz-date"
      case xAmzSecurityToken = "x-amz-security-token"
      case policy
      case xAmzSignature = "x-amz-signature"
    }
  }
  let url: String
  let fields: Fields
}

struct SingleUploadURLResponseBody: Decodable {
  let mediaId: String
  let url: URLData
}

struct SingleUploadURLResponse: Decodable {
  let body: SingleUploadURLResponseBody
  let statusCode: Int
}

class RegularUploader : IUploader {
  var type = "Single-shot"
  let presignedURL: String
  
  init(presignedURL: String) {
    self.presignedURL = presignedURL
  }
  // Default chunk size for determining single vs multipart upload
  static let DEFAULT_CHUNK_SIZE = Double(5 * 1024 * 1024)
  
  /// Returns the number of chunks for the file at the given path.
  func getNumChunks(path: String, chunkSize: Double = DEFAULT_CHUNK_SIZE) -> Int {
    guard let attributes = try? FileManager.default.attributesOfItem(atPath: path),
          let fileSize = attributes[.size] as? UInt64 else {
      return -1
    }
    return Int(ceil(Double(fileSize) / chunkSize))
  }
  
  /// For this example, use regular (non-multipart) upload if there is exactly one chunk.
  func useRegularUpload(numChunks: Int) -> Bool {
    return numChunks == 1
  }
  
  /// Regular (single-shot) file upload.
  func upload(token: String, path: String) async throws -> String {
    let fileURL = URL(fileURLWithPath: path)
    guard FileManager.default.fileExists(atPath: fileURL.path) else {
      throw NSError(domain: "RegularFileUploader", code: 404, userInfo: [NSLocalizedDescriptionKey: "File not found: \(path)"])
    }
    
    var request = URLRequest(url: URL(string: presignedURL)!)
    request.httpMethod = "POST"
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    let (data, response) = try await URLSession.shared.data(for: request)
    guard let httpResponse = response as? HTTPURLResponse, (200..<300).contains(httpResponse.statusCode) else {
      throw NSError(domain: "RegularFileUploader", code: 0, userInfo: [NSLocalizedDescriptionKey: "Fetching presigned URL failed"])
    }
    
    let presignedResponse = try JSONDecoder().decode(SingleUploadURLResponse.self, from: data)
    let mediaId = presignedResponse.body.mediaId
    let urlData = presignedResponse.body.url
    
    let boundary = "Boundary-\(UUID().uuidString)"
    // This is called multipart because it's a form-data object, not because it's meant to be a multipart upload
    var multipartRequest = URLRequest(url: URL(string: urlData.url)!)
    multipartRequest.httpMethod = "POST"
    multipartRequest.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
    
    var body = Data()
    let formFields: [(name: String, value: String)] = [
      ("key", urlData.fields.key),
      ("x-amz-algorithm", urlData.fields.xAmzAlgorithm),
      ("x-amz-credential", urlData.fields.xAmzCredential),
      ("x-amz-date", urlData.fields.xAmzDate),
      ("x-amz-security-token", urlData.fields.xAmzSecurityToken),
      ("policy", urlData.fields.policy),
      ("x-amz-signature", urlData.fields.xAmzSignature)
    ]
    
    // Add each field.
    for (name, value) in formFields {
      body.append("--\(boundary)\r\n".data(using: .utf8)!)
      body.append("Content-Disposition: form-data; name=\"\(name)\"\r\n\r\n".data(using: .utf8)!)
      body.append("\(value)\r\n".data(using: .utf8)!)
    }
    
    // Append file.
    let fileData = try Data(contentsOf: fileURL)
    body.append("--\(boundary)\r\n".data(using: .utf8)!)
    let filename = fileURL.lastPathComponent
    body.append("Content-Disposition: form-data; name=\"file\"; filename=\"\(filename)\"\r\n".data(using: .utf8)!)
    body.append("Content-Type: application/octet-stream\r\n\r\n".data(using: .utf8)!)
    body.append(fileData)
    body.append("\r\n".data(using: .utf8)!)
    body.append("--\(boundary)--\r\n".data(using: .utf8)!)
    
    multipartRequest.httpBody = body
    let (_, uploadResponse) = try await URLSession.shared.data(for: multipartRequest)
    guard let uploadHTTPResponse = uploadResponse as? HTTPURLResponse, (200..<300).contains(uploadHTTPResponse.statusCode) else {
      throw NSError(domain: "RegularFileUploader", code: 0, userInfo: [NSLocalizedDescriptionKey: "File upload failed"])
    }
    return mediaId
  }
}
